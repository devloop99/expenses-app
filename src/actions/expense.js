import database from "../firebase/firebase-101";

export const addExpense = (expense = {}) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expense) => {
  return (dispatch) => {
    const { description = "", note = "", amount = 0, createdAt = 0 } = expense;
    expense = { description, amount, note, createdAt };
    return database
      .ref("expenses")
      .push(expense)
      .then((ref) => {
        dispatch(addExpense({ id: ref.key, ...expense }));
      });
  };
};
export const removeExpense = ({ id }) => {
  return {
    type: "REMOVE_EXPENSE",
    id,
  };
};

export const startRemoveExpense = ({ id }) => {
  console.log("im called");
  return (dispatch) => {
    database
      .ref(`expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
  };
};

export const editExpense = (expense, id) => ({
  type: "EDIT_EXPENSE",
  id,
  expense,
});

export const startEditExpense = (expense, id) => {
  return (dispatch) => {
    console.log(expense);

    database
      .ref(`expenses/${id}`)
      .update(expense)
      .then(() => {
        dispatch(editExpense(expense, id));
      });
  };
};
