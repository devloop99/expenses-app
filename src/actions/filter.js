export const setText = (text = "") => ({
  type: "SET_TEXT",
  text,
});
export const sortByDate = () => ({
  type: "SORT_BY_DATE",
});
export const sortByAmount = () => ({
  type: "SORT_BY_AMOUNT",
});

export const setStartDate = (stamps = undefined) => ({
  type: "START_DATE",
  startDate: stamps,
});

export const setEndDate = (stamps = undefined) => ({
  type: "END_DATE",
  startDate: stamps,
});
