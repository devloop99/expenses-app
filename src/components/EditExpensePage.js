import React from "react";
import ExpenseForme from "./ExpenseForme";
import { connect } from "react-redux";
import {
  startEditExpense as editExpense,
  startRemoveExpense as removeExpense,
} from "../actions/expense";

class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      expense: props.expenses.find((el) => el.id === props.match.params.id),
    };
  }
  onSubmit = (expense) => {
    const id = this.props.match.params.id;

    this.props.dispatch(editExpense(expense, id));
    this.props.history.push("/");
  };
  onRemove = () => {
    console.log("loading rm ....");
    console.log(this.props.match.params.id);

    this.props.dispatch(removeExpense({ id: this.props.match.params.id }));
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h2>EDITTING EXPENSE :</h2>
        <ExpenseForme expense={this.state.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onRemove}>remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
  };
};
const EditExpensePageConnected = connect(mapStateToProps)(EditExpensePage);

export default EditExpensePageConnected;
