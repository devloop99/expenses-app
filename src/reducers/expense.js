const defaultState = {
  expenses: [],
  filters: {
    text: "",
    description: "",
    note: "",
    createdAt: undefined,
  },
};

const expenseReducer = (state = [], action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);

    case "EDIT_EXPENSE":
      return state.map((el) => {
        if (el.id === action.id) {
          return {
            ...el,
            ...action.expense,
          };
        }

        return el;
      });

    default:
      return state;
  }
};

export default expenseReducer;
