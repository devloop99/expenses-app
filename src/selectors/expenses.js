const getVisible = (expense, { text, sortBy, startDate, endDate }) => {
  const visible = expense
    .filter((el) => {
      const startMatch =
        typeof startDate !== "number" || el.createdAt >= startDate;

      const endMatch = typeof endDate !== "number" || el.createdAt <= endDate;
      const textMatch =
        el.description.toLowerCase().includes(text) || el.note.includes(text);

      return startMatch && endMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return -(a.createdAt - b.createdAt);
      } else if (sortBy === "amount") {
        return b.amount - a.amount;
      }
    });

  return visible;
};

export default getVisible;
