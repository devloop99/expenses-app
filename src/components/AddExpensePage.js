import React from "react";
import ExpenseForme from "./ExpenseForme";
import { connect } from "react-redux";
import { startAddExpense } from "../actions/expense";

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h2>CREATE EXPENSE :</h2>
        <ExpenseForme onSubmit={this.onSubmit} />
      </div>
    );
  }
}
const mapDispatchToProp = (dispatch) => ({
  addExpense: (expense) => dispatch(startAddExpense(expense)),
});
const AddExpensePageConnected = connect(
  undefined,
  mapDispatchToProp
)(AddExpensePage);

export default AddExpensePageConnected;
