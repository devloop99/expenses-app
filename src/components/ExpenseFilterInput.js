import React from "react";
import { connect } from "react-redux";
import { setText, SortByAmount, SortByDate } from "../actions/filter";
import { DateRangePicker } from "react-dates";
import moment from "moment";
import { setEndDate, setStartDate } from "../actions/filter";

class ExpenseFilterInput extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      option: "date",
      isFocused: undefined,
      startDate: moment().startOf("month"),
      endDate: moment().endOf("month"),
    };
  }
  changeHandler = (e) => {
    const option = e.target.value;
    this.setState(() => ({ option }));
    if (option === "date") {
      this.props.dispatch(SortByDate());
    } else if (option === "amount") {
      this.props.dispatch(SortByAmount());
    }
  };
  dateChange = (focusedInput) => {
    this.setState({ isFocused: focusedInput });
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={(e) => {
            this.props.dispatch(setText(e.target.value));
          }}
        />
        <select value={this.state.option} onChange={this.changeHandler}>
          <option value="amount">Amount</option>
          <option value="date">Date</option>
        </select>
        <DateRangePicker
          startDate={this.state.startDate} // momentPropTypes.momentObj or null,
          endDate={this.state.endDate} // momentPropTypes.momentObj or null,
          endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
          onDatesChange={({ startDate, endDate }) => {
            this.setState(() => ({ startDate }));
            this.setState(() => ({ endDate }));
            this.props.dispatch(
              setStartDate(startDate ? startDate.valueOf() : 0)
            );
            this.props.dispatch(
              setEndDate(endDate ? endDate.valueOf() : undefined)
            );
          }} // PropTypes.func.isRequired,
          focusedInput={this.state.isFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
          onFocusChange={this.dateChange} // PropTypes.func.isRequired,
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

const ExpenseFilterInputConnected = connect(mapStateToProps)(
  ExpenseFilterInput
);

export default ExpenseFilterInputConnected;
