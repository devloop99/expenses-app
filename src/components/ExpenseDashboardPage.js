import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseFilterInput from "./ExpenseFilterInput";
import ExpenseTotal from "./ExpenseTotal";

export const ExpenseDashboardPage = () => (
  <div>
    <ExpenseTotal />
    <ExpenseFilterInput />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
