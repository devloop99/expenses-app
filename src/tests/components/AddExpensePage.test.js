import React from "react";
import { AddExpensePage } from "../../components/AddExpensePage";
import { shallow } from "enzyme";
import { startAddExpense } from "../../actions/expense";
import expenses from "../fixtures/expenses";

test("should add expense page render correctly", () => {
  const addExpense = jest.fn();
  const history = { push: jest.fn() };
  const wrapper = shallow(
    <AddExpensePage history={history} addExpense={addExpense} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should handle submit", () => {
  const addExpense = jest.fn();
  const history = { push: jest.fn() };
  const wrapper = shallow(
    <AddExpensePage history={history} addExpense={addExpense} />
  );
  wrapper.find("ExpenseForme").prop("onSubmit")(expenses[1]);
  expect(addExpense).toHaveBeenLastCalledWith(expenses[1]);
});
