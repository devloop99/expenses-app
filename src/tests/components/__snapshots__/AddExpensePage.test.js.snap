// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add expense page render correctly 1`] = `
<div>
  <h2>
    CREATE EXPENSE :
  </h2>
  <ExpenseForme
    onSubmit={[Function]}
  />
</div>
`;
